src/MagnumPlugins/UfbxImporter/UfbxImporter.cpp

```diff
@@ -0,0 +1,1013 @@
+    { MaterialAttributeType::Vector4, MaterialAttribute::BaseColor, MaterialAttribute::BaseColorTexture, MaterialAttribute::BaseColorTextureMatrix, MaterialAttribute::BaseColorTextureCoordinates, UFBX_MATERIAL_PBR_BASE_COLOR, UFBX_MATERIAL_PBR_BASE_FACTOR },
+    { MaterialAttributeType::Float, MaterialAttribute::Metalness, MaterialAttribute::MetalnessTexture, MaterialAttribute::MetalnessTextureMatrix, MaterialAttribute::MetalnessTextureCoordinates, UFBX_MATERIAL_PBR_METALNESS, -1 },
+    { MaterialAttributeType::Float, MaterialAttribute::Roughness, MaterialAttribute::RoughnessTexture, MaterialAttribute::RoughnessTextureMatrix, MaterialAttribute::RoughnessTextureCoordinates, UFBX_MATERIAL_PBR_ROUGHNESS, -1 },
+    { MaterialAttributeType::Vector3, MaterialAttribute::EmissiveColor, MaterialAttribute::EmissiveTexture, MaterialAttribute::EmissiveTextureMatrix, MaterialAttribute::EmissiveTextureCoordinates, UFBX_MATERIAL_PBR_EMISSION_COLOR, UFBX_MATERIAL_PBR_EMISSION_FACTOR },
+    { (MaterialAttributeType)0, (MaterialAttribute)0, MaterialAttribute::NormalTexture, MaterialAttribute::NormalTextureMatrix, MaterialAttribute::NormalTextureCoordinates, UFBX_MATERIAL_PBR_NORMAL_MAP, -1 },
+    { (MaterialAttributeType)0, (MaterialAttribute)0, MaterialAttribute::OcclusionTexture, MaterialAttribute::OcclusionTextureMatrix, MaterialAttribute::OcclusionTextureCoordinates, UFBX_MATERIAL_PBR_AMBIENT_OCCLUSION, -1 },
+};
+
+}
+
+struct TextureBlob {
+    const void *data;
+    std::size_t size;
+};
+
+struct UfbxImporter::State {
+    ufbx_scene_ref scene;
+
+    Containers::Array<MeshChunk> meshChunks;
+
+    /* Mapping from ufbx_mesh.typed_id -> State.meshChunks */
+    Containers::Array<UnsignedInt> meshChunkBase;
+
+    UnsignedInt nodeIdOffset = 0;
+    UnsignedInt originalNodeCount = 0;
+    UnsignedInt nodeCountWithSynthetic = 0;
+
+    bool fromFile = false;
+
+    UnsignedInt imageImporterId = ~UnsignedInt{};
+    Containers::Optional<AnyImageImporter> imageImporter;
+};
+
+UfbxImporter::UfbxImporter() {
+}
+
+UfbxImporter::UfbxImporter(PluginManager::AbstractManager& manager, const Containers::StringView& plugin): AbstractImporter{manager, plugin} {
+}
+
+UfbxImporter::~UfbxImporter() = default;
+
+ImporterFeatures UfbxImporter::doFeatures() const { return ImporterFeature::OpenData|ImporterFeature::OpenState|ImporterFeature::FileCallback; }
+
+bool UfbxImporter::doIsOpened() const { return !!_state; }
+
+void UfbxImporter::doClose() { _state = nullptr; }
+
+void UfbxImporter::doOpenData(Containers::Array<char>&& data, const DataFlags) {
+    ufbx_load_opts opts = loadOptsFromConfiguration(configuration());
+    ufbx_error error;
+    ufbx_scene *scene = ufbx_load_memory(data.data(), data.size(), &opts, &error);
+    if (!scene) {
+        logError("Trade::UfbxImporter::openData(): loading failed: ", error);
+    }
+
+    /* Handle the opened scene with doOpenState() */
+    doOpenState(scene, {});
+    ufbx_free(scene);
```

Comment:
I think this is freeing a bit early. 

Just to clarify, the interface is done with lazy loading in mind -- where `openData()` would do just some minimal initial setup, and then the various `doMesh()` etc functions are called only once the user requests given piece of data, and are doing the heavy lifting with data copying, conversion etc. It's not like the `doMesh()` etc APIs would be called all upfront as part of `openData()`, the results stored somewhere and then just given to users.

Thus the state is meant to be alive until the file closed again, so `ufbx_free(scene)` should be in `doClose()` instead, and the destructor as well. Or, for a more hermetic code, have a destructor on the `State` itself:

```cpp
struct UfbxImporter::State {
    ~State() {
        ufbx_free(scene);
    }

    ufbx_scene_ref scene{};
    ...
```

I feel this could also maybe resolve the strange issues you had with dynamic plugins?

Reply:
This is again with my confusing `ufbx_scene_ref`, maybe I should move it to some C++ utility thing. I tend to prototype new APIs when writing user code as then you tend to end up with "what the user wants" instead of "what the library writer decides might be useful". Internally `ufbx_scene` and friends are refcounted and `doOpenState()` increments it using `ufbx_retain()`, the idea was that if an user passes a scene in the importer doesn't take ownership of it.

Reply:
Ahh, I see now, thanks. I got extremely confused because, seeing a `.c` source, I didn't expect any magical RAII things to be happening. Everything is clear now :)